{
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "# Agent 3\n",
        "\n",
        "The outputs generated by Agents may contain technical details or extraneous information, making them redundant or meaningless to users. To enhance readability, we developed Agent 3, which refines and restructures Python output strings into standardized, user-friendly formats. For example, Python lists are reformatted as clear, numbered lists, and file download information is converted into simple, intuitive messages. This ensures that responses are concise, relevant, and easy for users to understand, regardless of the original technical formatting."
      ],
      "metadata": {
        "id": "YCi0SWp3TnHl"
      },
      "id": "YCi0SWp3TnHl"
    },
    {
      "cell_type": "code",
      "source": [
        "import re, io, sys"
      ],
      "metadata": {
        "id": "_qEZe_kiZYvd",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 17
        },
        "executionInfo": {
          "status": "ok",
          "timestamp": 1731092491152,
          "user_tz": 300,
          "elapsed": 359,
          "user": {
            "displayName": "",
            "userId": ""
          }
        },
        "outputId": "b82e5d42-6209-490c-d50f-9421bd074ff3"
      },
      "id": "_qEZe_kiZYvd",
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "\n",
              "  <style>\n",
              "    pre {\n",
              "        white-space: pre-wrap;\n",
              "    }\n",
              "  </style>\n",
              "  "
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from IPython.display import HTML, display, Markdown\n",
        "\n",
        "def set_css():\n",
        "  display(HTML('''\n",
        "  <style>\n",
        "    pre {\n",
        "        white-space: pre-wrap;\n",
        "    }\n",
        "  </style>\n",
        "  '''))\n",
        "get_ipython().events.register('pre_run_cell', set_css)"
      ],
      "metadata": {
        "id": "Hxh9c5cjVU8U",
        "executionInfo": {
          "status": "ok",
          "timestamp": 1731092488585,
          "user_tz": 300,
          "elapsed": 5,
          "user": {
            "displayName": "",
            "userId": ""
          }
        }
      },
      "id": "Hxh9c5cjVU8U",
      "execution_count": 4,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import base64\n",
        "import vertexai\n",
        "from vertexai.generative_models import GenerativeModel, Part, SafetySetting"
      ],
      "metadata": {
        "id": "Te7k0F13EICY",
        "executionInfo": {
          "status": "ok",
          "timestamp": 1731092461506,
          "user_tz": 300,
          "elapsed": 2203,
          "user": {
            "displayName": "",
            "userId": ""
          }
        }
      },
      "id": "Te7k0F13EICY",
      "execution_count": 3,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "PROJECT_ID = \"isb-cgc-external-004\"\n",
        "LOCATION = \"us-central1\"\n",
        "\n",
        "import vertexai\n",
        "vertexai.init(project=PROJECT_ID, location=LOCATION)"
      ],
      "metadata": {
        "id": "d9kb9xRPEX1N",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 17
        },
        "executionInfo": {
          "status": "ok",
          "timestamp": 1731092495955,
          "user_tz": 300,
          "elapsed": 281,
          "user": {
            "displayName": "",
            "userId": ""
          }
        },
        "outputId": "d9e82e7f-dfdd-4a69-fb8b-cf8bf8d881a6"
      },
      "id": "d9kb9xRPEX1N",
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "\n",
              "  <style>\n",
              "    pre {\n",
              "        white-space: pre-wrap;\n",
              "    }\n",
              "  </style>\n",
              "  "
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "generation_config = {\n",
        "    \"max_output_tokens\": 8192,\n",
        "    \"temperature\": 1,\n",
        "    \"top_p\": 0.5,\n",
        "}\n",
        "\n",
        "safety_settings = [\n",
        "    SafetySetting(\n",
        "        category=SafetySetting.HarmCategory.HARM_CATEGORY_HATE_SPEECH,\n",
        "        threshold=SafetySetting.HarmBlockThreshold.OFF\n",
        "    ),\n",
        "    SafetySetting(\n",
        "        category=SafetySetting.HarmCategory.HARM_CATEGORY_DANGEROUS_CONTENT,\n",
        "        threshold=SafetySetting.HarmBlockThreshold.OFF\n",
        "    ),\n",
        "    SafetySetting(\n",
        "        category=SafetySetting.HarmCategory.HARM_CATEGORY_SEXUALLY_EXPLICIT,\n",
        "        threshold=SafetySetting.HarmBlockThreshold.OFF\n",
        "    ),\n",
        "    SafetySetting(\n",
        "        category=SafetySetting.HarmCategory.HARM_CATEGORY_HARASSMENT,\n",
        "        threshold=SafetySetting.HarmBlockThreshold.OFF\n",
        "    ),\n",
        "]\n",
        "\n",
        "def generate(question = \"\"\"What can you do?\"\"\"):\n",
        "\n",
        "    ## General prompt\n",
        "    prompt = \"\"\"You are an experienced writer and editor. When given an input, convert it into a well formatted string output. If the input string is blank, just say that there are no outputs generated for the current query. When someone ask you who you are, return accordingly.\n",
        "\n",
        "    Here are some examples:\n",
        "    Example 1:\n",
        "    Input: '''\n",
        "    Generated code: import scanpy as sc\n",
        "    import pandas as pd\n",
        "\n",
        "    adata = sc.read_h5ad(\"/content/demo_data/6723_KL_1_unfiltered.h5ad\")\n",
        "    print(adata)\n",
        "    AnnData object with n_obs × n_vars = 4992 × 19074\n",
        "    obs: 'in_tissue', 'array_row', 'array_col', 'kmeans_7_clusters', 'kmeans_10_clusters', 'kmeans_4_clusters', 'kmeans_2_clusters', 'kmeans_6_clusters', 'graphclust', 'kmeans_3_clusters', 'kmeans_8_clusters', 'kmeans_9_clusters', 'kmeans_5_clusters'\n",
        "    var: 'gene_ids', 'feature_types', 'genome', 'n_cells', 'Morans_I', 'Morans_I_p_val', 'Morans_I_adj_p_val', 'Feature Counts in Spots Under Tissue', 'Median Normalized Average Counts', 'Barcodes Detected per Feature'\n",
        "    uns: 'spatial'\n",
        "    obsm: 'spatial'\n",
        "    Agent 2: Executed Python code with accessible variables.\n",
        "    The agent action is tool='agent_2' tool_input={'question': 'Can you use scanpy to load the /content/demo_data/6723_KL_1_unfiltered.h5ad?'} log=\"\\nInvoking: `agent_2` with `{'question': 'Can you use scanpy to load the /content/demo_data/6723_KL_1_unfiltered.h5ad?'}`\\n\\n\\n\" message_log=[AIMessage(content='', additional_kwargs={'function_call': {'name': 'agent_2', 'arguments': '{\"question\": \"Can you use scanpy to load the /content/demo_data/6723_KL_1_unfiltered.h5ad?\"}'}}, response_metadata={'is_blocked': False, 'safety_ratings': [], 'usage_metadata': {'prompt_token_count': 380, 'candidates_token_count': 35, 'total_token_count': 415, 'cached_content_token_count': 0}, 'finish_reason': 'STOP', 'avg_logprobs': -0.004535583513123649, 'logprobs_result': {'top_candidates': [], 'chosen_candidates': []}}, id='run-7d91d738-3af2-468a-bb2b-e9b23235e38a-0', tool_calls=[{'name': 'agent_2', 'args': {'question': 'Can you use scanpy to load the /content/demo_data/6723_KL_1_unfiltered.h5ad?'}, 'id': '62e66320-2760-4403-9049-43fd7a676e0f', 'type': 'tool_call'}], usage_metadata={'input_tokens': 380, 'output_tokens': 35, 'total_tokens': 415})] tool_call_id='62e66320-2760-4403-9049-43fd7a676e0f'\n",
        "    The tool result is: None\n",
        "    {'intermediate_steps': [(ToolAgentAction(tool='agent_2', tool_input={'question': 'Can you use scanpy to load the /content/demo_data/6723_KL_1_unfiltered.h5ad?'}, log=\"\\nInvoking: `agent_2` with `{'question': 'Can you use scanpy to load the /content/demo_data/6723_KL_1_unfiltered.h5ad?'}`\\n\\n\\n\", message_log=[AIMessage(content='', additional_kwargs={'function_call': {'name': 'agent_2', 'arguments': '{\"question\": \"Can you use scanpy to load the /content/demo_data/6723_KL_1_unfiltered.h5ad?\"}'}}, response_metadata={'is_blocked': False, 'safety_ratings': [], 'usage_metadata': {'prompt_token_count': 380, 'candidates_token_count': 35, 'total_token_count': 415, 'cached_content_token_count': 0}, 'finish_reason': 'STOP', 'avg_logprobs': -0.004535583513123649, 'logprobs_result': {'top_candidates': [], 'chosen_candidates': []}}, id='run-7d91d738-3af2-468a-bb2b-e9b23235e38a-0', tool_calls=[{'name': 'agent_2', 'args': {'question': 'Can you use scanpy to load the /content/demo_data/6723_KL_1_unfiltered.h5ad?'}, 'id': '62e66320-2760-4403-9049-43fd7a676e0f', 'type': 'tool_call'}], usage_metadata={'input_tokens': 380, 'output_tokens': 35, 'total_tokens': 415})], tool_call_id='62e66320-2760-4403-9049-43fd7a676e0f'), 'None')]}\n",
        "\n",
        "    '''\n",
        "    Well formated output:\n",
        "\n",
        "    '''The dataset is loaded to the anndata object named adata. There are 4992 cells or spots, 19074 genes, and 13 feature columns. The data object is printed below:\n",
        "\n",
        "    AnnData object with n_obs × n_vars = 4992 × 19074\n",
        "    obs: 'in_tissue', 'array_row', 'array_col', 'kmeans_7_clusters', 'kmeans_10_clusters', 'kmeans_4_clusters', 'kmeans_2_clusters', 'kmeans_6_clusters', 'graphclust', 'kmeans_3_clusters', 'kmeans_8_clusters', 'kmeans_9_clusters', 'kmeans_5_clusters'\n",
        "    var: 'gene_ids', 'feature_types', 'genome', 'n_cells', 'Morans_I', 'Morans_I_p_val', 'Morans_I_adj_p_val', 'Feature Counts in Spots Under Tissue', 'Median Normalized Average Counts', 'Barcodes Detected per Feature'\n",
        "    uns: 'spatial'\n",
        "    obsm: 'spatial'\n",
        "    '''\n",
        "\n",
        "    Example 2:\n",
        "    Input: '''\n",
        "    Welcome, arun.das!\n",
        "\n",
        "    INFO:synapseclient_default:Welcome, arun.das!\n",
        "\n",
        "    Downloading files:  77%|███████▋  | 16.8M/21.8M [00:00<00:00, 28.0MB/s, syn51133599]Downloaded syn51133599 to /content/demo_data/8578_AS_1_unfiltered.h5ad\n",
        "    [INFO] Downloaded syn51133599 to /content/demo_data/8578_AS_1_unfiltered.h5ad\n",
        "    Downloading files: 100%|██████████| 21.8M/21.8M [00:00<00:00, 31.8MB/s, syn51133599]INFO:synapseclient_default:Downloaded syn51133599 to /content/demo_data/8578_AS_1_unfiltered.h5ad\n",
        "    Downloading files: 100%|██████████| 21.8M/21.8M [00:00<00:00, 31.6MB/s, syn51133599]import synapseclient\n",
        "    syn = synapseclient.login(silent=True)\n",
        "    entity = syn.get('syn51133599', downloadLocation='/content/demo_data')\n",
        "    The agent action is tool='agent_1' tool_input={'question': 'Can you download the synapse dataset syn51133599 to /content/demo_data/?'} log=\"\\nInvoking: `agent_1` with `{'question': 'Can you download the synapse dataset syn51133599 to /content/demo_data/?'}`\\n\\n\\n\" message_log=[AIMessage(content='', additional_kwargs={'function_call': {'name': 'agent_1', 'arguments': '{\"question\": \"Can you download the synapse dataset syn51133599 to /content/demo_data/?\"}'}}, response_metadata={'is_blocked': False, 'safety_ratings': [], 'usage_metadata': {'prompt_token_count': 396, 'candidates_token_count': 27, 'total_token_count': 423, 'cached_content_token_count': 0}, 'finish_reason': 'STOP', 'avg_logprobs': -0.0003545089038433852, 'logprobs_result': {'top_candidates': [], 'chosen_candidates': []}}, id='run-49c29b88-a638-4a7d-89fa-122d8edb910d-0', tool_calls=[{'name': 'agent_1', 'args': {'question': 'Can you download the synapse dataset syn51133599 to /content/demo_data/?'}, 'id': '48d647d4-94d2-4cae-bd6f-f5d5e2240945', 'type': 'tool_call'}], usage_metadata={'input_tokens': 396, 'output_tokens': 27, 'total_tokens': 423})] tool_call_id='48d647d4-94d2-4cae-bd6f-f5d5e2240945'\n",
        "    The tool result is: None\n",
        "    ----\n",
        "    '''\n",
        "\n",
        "    Well formated output:\n",
        "    '''\n",
        "    The synapse dataset with synapse id syn51133599 is successfully downloaded to /content/demo_data/8578_AS_1_unfiltered.h5ad.\n",
        "    '''\n",
        "\n",
        "\n",
        "    Example 3:\n",
        "    Input: '''\n",
        "    {'agent_outcome': [ToolAgentAction(tool='agent_1', tool_input={'question': 'Can you load the entityId of all data under HTAN.10xvisium_spatialtranscriptomics_scRNAseq_level4_metadata_current where the File_Format is hdf5?'}, log=\"\\nInvoking: `agent_1` with `{'question': 'Can you load the entityId of all data under HTAN.10xvisium_spatialtranscriptomics_scRNAseq_level4_metadata_current where the File_Format is hdf5?'}`\\n\\n\\n\", message_log=[AIMessage(content='', additional_kwargs={'function_call': {'name': 'agent_1', 'arguments': '{\"question\": \"Can you load the entityId of all data under HTAN.10xvisium_spatialtranscriptomics_scRNAseq_level4_metadata_current where the File_Format is hdf5?\"}'}}, response_metadata={'is_blocked': False, 'safety_ratings': [], 'usage_metadata': {'prompt_token_count': 536, 'candidates_token_count': 49, 'total_token_count': 585, 'cached_content_token_count': 0}, 'finish_reason': 'STOP', 'avg_logprobs': -0.0002904396352111077, 'logprobs_result': {'top_candidates': [], 'chosen_candidates': []}}, id='run-b2cf906b-d897-4164-b743-1120419ebdf9-0', tool_calls=[{'name': 'agent_1', 'args': {'question': 'Can you load the entityId of all data under HTAN.10xvisium_spatialtranscriptomics_scRNAseq_level4_metadata_current where the File_Format is hdf5?'}, 'id': '6987b536-4024-4b91-a0b1-720e41b1076e', 'type': 'tool_call'}], usage_metadata={'input_tokens': 536, 'output_tokens': 49, 'total_tokens': 585})], tool_call_id='6987b536-4024-4b91-a0b1-720e41b1076e')]}\n",
        "    import pandas_gbq\n",
        "    from google.cloud import bigquery\n",
        "    import synapseclient\n",
        "\n",
        "    project_id = \"isb-cgc-external-004\"\n",
        "\n",
        "    query = '''\n",
        "    SELECT entityId\n",
        "    FROM `isb-cgc-bq.HTAN.10xvisium_spatialtranscriptomics_scRNAseq_level4_metadata_current`\n",
        "    WHERE File_Format = 'hdf5'\n",
        "    '''\n",
        "\n",
        "    df = pandas_gbq.read_gbq(query, project_id=project_id)\n",
        "    print(df)\n",
        "\n",
        "    syn = synapseclient.login(silent=True)\n",
        "    entity = syn.get('syn51133602', downloadLocation='/content/datasets')\n",
        "    Downloading: 100%|##########|\n",
        "        entityId\n",
        "    0   syn51133519\n",
        "    1   syn51133520\n",
        "    2   syn51133521\n",
        "    3   syn51133522\n",
        "    4   syn51133523\n",
        "    5   syn51133524\n",
        "    6   syn51133525\n",
        "    7   syn51133526\n",
        "    8   syn51133527\n",
        "    9   syn51133528\n",
        "    10  syn51133529\n",
        "    11  syn51133530\n",
        "    12  syn51133531\n",
        "    13  syn51133532\n",
        "    14  syn51133533\n",
        "    15  syn51133534\n",
        "    16  syn51133537\n",
        "    17  syn51133540\n",
        "    18  syn51133578\n",
        "    19  syn51133580\n",
        "    20  syn51133583\n",
        "    21  syn51133587\n",
        "    22  syn51133591\n",
        "    23  syn51133592\n",
        "    24  syn51133593\n",
        "    25  syn51133595\n",
        "    26  syn51133596\n",
        "    27  syn51133597\n",
        "    28  syn51133598\n",
        "    29  syn51133599\n",
        "    30  syn51133600\n",
        "    31  syn51133601\n",
        "    32  syn51133602\n",
        "    33  syn51133603\n",
        "    34  syn51133604\n",
        "    35  syn51133605\n",
        "    36  syn51133606\n",
        "    37  syn51133607\n",
        "    38  syn51133608\n",
        "    39  syn51133609\n",
        "    40  syn51133612\n",
        "    The agent action is tool='agent_1' tool_input={'question': 'Can you load the entityId of all data under HTAN.10xvisium_spatialtranscriptomics_scRNAseq_level4_metadata_current where the File_Format is hdf5?'} log=\"\\nInvoking: `agent_1` with `{'question': 'Can you load the entityId of all data under HTAN.10xvisium_spatialtranscriptomics_scRNAseq_level4_metadata_current where the File_Format is hdf5?'}`\\n\\n\\n\" message_log=[AIMessage(content='', additional_kwargs={'function_call': {'name': 'agent_1', 'arguments': '{\"question\": \"Can you load the entityId of all data under HTAN.10xvisium_spatialtranscriptomics_scRNAseq_level4_metadata_current where the File_Format is hdf5?\"}'}}, response_metadata={'is_blocked': False, 'safety_ratings': [], 'usage_metadata': {'prompt_token_count': 536, 'candidates_token_count': 49, 'total_token_count': 585, 'cached_content_token_count': 0}, 'finish_reason': 'STOP', 'avg_logprobs': -0.0002904396352111077, 'logprobs_result': {'top_candidates': [], 'chosen_candidates': []}}, id='run-b2cf906b-d897-4164-b743-1120419ebdf9-0', tool_calls=[{'name': 'agent_1', 'args': {'question': 'Can you load the entityId of all data under HTAN.10xvisium_spatialtranscriptomics_scRNAseq_level4_metadata_current where the File_Format is hdf5?'}, 'id': '6987b536-4024-4b91-a0b1-720e41b1076e', 'type': 'tool_call'}], usage_metadata={'input_tokens': 536, 'output_tokens': 49, 'total_tokens': 585})] tool_call_id='6987b536-4024-4b91-a0b1-720e41b1076e'\n",
        "    The tool result is: None\n",
        "    {'intermediate_steps': [(ToolAgentAction(tool='agent_1', tool_input={'question': 'Can you load the entityId of all data under HTAN.10xvisium_spatialtranscriptomics_scRNAseq_level4_metadata_current where the File_Format is hdf5?'}, log=\"\\nInvoking: `agent_1` with `{'question': 'Can you load the entityId of all data under HTAN.10xvisium_spatialtranscriptomics_scRNAseq_level4_metadata_current where the File_Format is hdf5?'}`\\n\\n\\n\", message_log=[AIMessage(content='', additional_kwargs={'function_call': {'name': 'agent_1', 'arguments': '{\"question\": \"Can you load the entityId of all data under HTAN.10xvisium_spatialtranscriptomics_scRNAseq_level4_metadata_current where the File_Format is hdf5?\"}'}}, response_metadata={'is_blocked': False, 'safety_ratings': [], 'usage_metadata': {'prompt_token_count': 536, 'candidates_token_count': 49, 'total_token_count': 585, 'cached_content_token_count': 0}, 'finish_reason': 'STOP', 'avg_logprobs': -0.0002904396352111077, 'logprobs_result': {'top_candidates': [], 'chosen_candidates': []}}, id='run-b2cf906b-d897-4164-b743-1120419ebdf9-0', tool_calls=[{'name': 'agent_1', 'args': {'question': 'Can you load the entityId of all data under HTAN.10xvisium_spatialtranscriptomics_scRNAseq_level4_metadata_current where the File_Format is hdf5?'}, 'id': '6987b536-4024-4b91-a0b1-720e41b1076e', 'type': 'tool_call'}], usage_metadata={'input_tokens': 536, 'output_tokens': 49, 'total_tokens': 585})], tool_call_id='6987b536-4024-4b91-a0b1-720e41b1076e'), 'None')]}\n",
        "    ---- Initial response captured ----\n",
        "    '''\n",
        "    Well formatted output:\n",
        "    '''\n",
        "    A total of 41 entity IDs were retrieved from the BigQuery table isb-cgc-bq.HTAN.10xvisium_spatialtranscriptomics_scRNAseq_level4_metadata_current where the File_Format is 'hdf5'.  The entity IDs are listed below::\n",
        "\n",
        "    1. syn51133519\n",
        "    2. syn51133520\n",
        "    3. syn51133521\n",
        "    4. syn51133522\n",
        "    5. syn51133523\n",
        "    6. syn51133524\n",
        "    7. syn51133525\n",
        "    8. syn51133526\n",
        "    9. syn51133527\n",
        "    10. syn51133528\n",
        "    11. syn51133529\n",
        "    12. syn51133530\n",
        "    13. syn51133531\n",
        "    14. syn51133532\n",
        "    15. syn51133533\n",
        "    16. syn51133534\n",
        "    17. syn51133537\n",
        "    18. syn51133540\n",
        "    19. syn51133578\n",
        "    20. syn51133580\n",
        "    21. syn51133583\n",
        "    22. syn51133587\n",
        "    23. syn51133591\n",
        "    24. syn51133592\n",
        "    25. syn51133593\n",
        "    26. syn51133595\n",
        "    27. syn51133596\n",
        "    28. syn51133597\n",
        "    29. syn51133598\n",
        "    30. syn51133599\n",
        "    31. syn51133600\n",
        "    32. syn51133601\n",
        "    33. syn51133602\n",
        "    34. syn51133603\n",
        "    35. syn51133604\n",
        "    36. syn51133605\n",
        "    37. syn51133606\n",
        "    38. syn51133607\n",
        "    39. syn51133608\n",
        "    40. syn51133609\n",
        "    41. syn51133612\n",
        "    '''\n",
        "    \"\"\"\n",
        "\n",
        "    prompt += f\"\"\"\n",
        "    Input: {question}\n",
        "    Please generate the well formated output.\n",
        "    \"\"\"\n",
        "\n",
        "    vertexai.init(project=\"isb-cgc-external-004\", location=\"us-central1\")\n",
        "    model = GenerativeModel(\n",
        "        \"gemini-1.5-flash-002\",\n",
        "    )\n",
        "    responses = model.generate_content(\n",
        "        [prompt],\n",
        "        generation_config=generation_config,\n",
        "        safety_settings=safety_settings,\n",
        "        stream=False,\n",
        "        # tools='code_execution'\n",
        "    )\n",
        "\n",
        "    # Example usage\n",
        "    display(Markdown(responses.text))\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 17
        },
        "id": "qhnOrkXrYMxQ",
        "executionInfo": {
          "status": "ok",
          "timestamp": 1731092554836,
          "user_tz": 300,
          "elapsed": 154,
          "user": {
            "displayName": "",
            "userId": ""
          }
        },
        "outputId": "d751da17-8e0b-4d70-84e0-8232a6522f04"
      },
      "id": "qhnOrkXrYMxQ",
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "\n",
              "  <style>\n",
              "    pre {\n",
              "        white-space: pre-wrap;\n",
              "    }\n",
              "  </style>\n",
              "  "
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "generate('Who are you and what can you do?')"
      ],
      "metadata": {
        "id": "NmYUZIiLKSFS"
      },
      "id": "NmYUZIiLKSFS",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Test 1\n",
        "\n",
        "text = \"\"\"\n",
        "candidates {\n",
        "  content {\n",
        "    role: \"model\"\n",
        "    parts {\n",
        "      text: \"The HTAN spatial transcriptomics metadata file contains information about the spatial location of cells, including attributes such as `Spatial Barcode Length`, `UMI Barcode Offset`, `UMI Barcode Length`, and potentially coordinates if available in the specific dataset.  It also includes general file information like `Filename`, `Run ID`, `File Format`, and HTAN identifiers such as `HTAN Parent Biospecimen ID` and `HTAN Data File ID`.  The exact attributes will vary depending on the specific dataset.\\n\"\n",
        "    }\n",
        "  }\n",
        "  finish_reason: STOP\n",
        "  avg_logprobs: -0.10189473395254098\n",
        "}\n",
        "usage_metadata {\n",
        "  prompt_token_count: 3155\n",
        "  candidates_token_count: 102\n",
        "  total_token_count: 3257\n",
        "}\n",
        "model_version: \"gemini-1.5-flash-002\"\n",
        "\"\"\"\n",
        "generate(text)"
      ],
      "metadata": {
        "id": "GO8pCZGoZRAi",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 104
        },
        "executionInfo": {
          "status": "ok",
          "timestamp": 1731092630855,
          "user_tz": 300,
          "elapsed": 1081,
          "user": {
            "displayName": "",
            "userId": ""
          }
        },
        "outputId": "50e1b929-83e8-459a-bb31-edc156353759"
      },
      "id": "GO8pCZGoZRAi",
      "execution_count": 11,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "\n",
              "  <style>\n",
              "    pre {\n",
              "        white-space: pre-wrap;\n",
              "    }\n",
              "  </style>\n",
              "  "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Markdown object>"
            ],
            "text/markdown": "```\nThe HTAN spatial transcriptomics metadata file contains information about the spatial location of cells, including attributes such as `Spatial Barcode Length`, `UMI Barcode Offset`, `UMI Barcode Length`, and potentially coordinates if available in the specific dataset. It also includes general file information like `Filename`, `Run ID`, `File Format`, and HTAN identifiers such as `HTAN Parent Biospecimen ID` and `HTAN Data File ID`. The exact attributes will vary depending on the specific dataset.\n```\n"
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "text = \"\"\"\n",
        "Downloading: 100%|██████████|\n",
        "         CellID cell_type\n",
        "0        271351     Other\n",
        "1        271535     Other\n",
        "2        271733     Other\n",
        "3        272229     Other\n",
        "4        272613     Other\n",
        "...         ...       ...\n",
        "6282743   83619     Other\n",
        "6282744  201761     Other\n",
        "6282745  450282     Other\n",
        "6282746   71082     Other\n",
        "6282747    5520     Other\n",
        "\n",
        "[6282748 rows x 2 columns]\n",
        "\"\"\"\n",
        "generate(text)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 64
        },
        "id": "UeL84jhjYxpm",
        "executionInfo": {
          "status": "ok",
          "timestamp": 1731092676820,
          "user_tz": 300,
          "elapsed": 795,
          "user": {
            "displayName": "",
            "userId": ""
          }
        },
        "outputId": "858775d4-8476-4cbb-98b0-f94d03c836d8"
      },
      "id": "UeL84jhjYxpm",
      "execution_count": 14,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "\n",
              "  <style>\n",
              "    pre {\n",
              "        white-space: pre-wrap;\n",
              "    }\n",
              "  </style>\n",
              "  "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Markdown object>"
            ],
            "text/markdown": "The provided input is a table showing cell IDs and their corresponding cell types.  There are 6,282,748 rows in the table.  The majority of cells are classified as \"Other\".  A more detailed analysis would require further information about the data and the desired level of detail in the output.\n"
          },
          "metadata": {}
        }
      ]
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "Python 3",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.10.10"
    },
    "colab": {
      "provenance": [],
      "name": "agent3_das.ipynb"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}