{
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "# Agent 1\n",
        "\n",
        "Agent 1 focuses on downloading the data / feed information to train the LLM to be able to handle queries related to HTAN spatial transcriptomics data using promt engineering. The prompts include\n",
        "* Example queries from HTAN community notebook\n",
        "* Downloading HTAN spatial datasets from Synapse database. The downloaded files will be used by Agent 2.\n",
        "  The information was obtained from\n",
        "this table `HTAN.10xvisium_spatialtranscriptomics_scRNAseq_level4_metadata_current`\n",
        "* Metadata information obtained from https://humantumoratlas.org/standard/spatial_transcriptomics\n",
        "\n",
        "Once the prompts have been provided, we then run some generalized queries to validate our work\n"
      ],
      "metadata": {
        "id": "YCi0SWp3TnHl"
      },
      "id": "YCi0SWp3TnHl"
    },
    {
      "cell_type": "code",
      "id": "qU5yhGERG5KXTudzh77T3qdR",
      "metadata": {
        "tags": [],
        "id": "qU5yhGERG5KXTudzh77T3qdR",
        "executionInfo": {
          "status": "ok",
          "timestamp": 1731012689416,
          "user_tz": 300,
          "elapsed": 3651,
          "user": {
            "displayName": "",
            "userId": ""
          }
        },
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 679
        },
        "outputId": "d4a4f4be-17e5-4dcb-a5e8-368e5b9b7bb0"
      },
      "source": [
        "!pip install synapseclient\n"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Requirement already satisfied: synapseclient in /usr/local/lib/python3.10/dist-packages (4.6.0)\n",
            "Requirement already satisfied: requests<3.0,>=2.22.0 in /usr/local/lib/python3.10/dist-packages (from synapseclient) (2.32.3)\n",
            "Requirement already satisfied: urllib3<2,>=1.26.18 in /usr/local/lib/python3.10/dist-packages (from synapseclient) (1.26.20)\n",
            "Requirement already satisfied: deprecated<2.0,>=1.2.4 in /usr/local/lib/python3.10/dist-packages (from synapseclient) (1.2.14)\n",
            "Requirement already satisfied: opentelemetry-api>=1.21.0 in /usr/local/lib/python3.10/dist-packages (from synapseclient) (1.28.0)\n",
            "Requirement already satisfied: opentelemetry-sdk>=1.21.0 in /usr/local/lib/python3.10/dist-packages (from synapseclient) (1.28.0)\n",
            "Requirement already satisfied: opentelemetry-exporter-otlp-proto-http>=1.21.0 in /usr/local/lib/python3.10/dist-packages (from synapseclient) (1.28.0)\n",
            "Requirement already satisfied: opentelemetry-instrumentation-httpx>=0.48b0 in /usr/local/lib/python3.10/dist-packages (from synapseclient) (0.49b0)\n",
            "Requirement already satisfied: opentelemetry-instrumentation-requests>=0.48b0 in /usr/local/lib/python3.10/dist-packages (from synapseclient) (0.49b0)\n",
            "Requirement already satisfied: opentelemetry-instrumentation-threading>=0.48b0 in /usr/local/lib/python3.10/dist-packages (from synapseclient) (0.49b0)\n",
            "Requirement already satisfied: opentelemetry-instrumentation-urllib>=0.48b0 in /usr/local/lib/python3.10/dist-packages (from synapseclient) (0.49b0)\n",
            "Requirement already satisfied: nest-asyncio~=1.6.0 in /usr/local/lib/python3.10/dist-packages (from synapseclient) (1.6.0)\n",
            "Requirement already satisfied: asyncio-atexit~=1.0.1 in /usr/local/lib/python3.10/dist-packages (from synapseclient) (1.0.1)\n",
            "Requirement already satisfied: httpx~=0.27.0 in /usr/local/lib/python3.10/dist-packages (from synapseclient) (0.27.2)\n",
            "Requirement already satisfied: tqdm<5.0,>=4.66.2 in /usr/local/lib/python3.10/dist-packages (from synapseclient) (4.66.5)\n",
            "Requirement already satisfied: async-lru~=2.0.4 in /usr/local/lib/python3.10/dist-packages (from synapseclient) (2.0.4)\n",
            "Requirement already satisfied: psutil~=5.9.8 in /usr/local/lib/python3.10/dist-packages (from synapseclient) (5.9.8)\n",
            "Requirement already satisfied: typing-extensions>=4.0.0 in /usr/local/lib/python3.10/dist-packages (from async-lru~=2.0.4->synapseclient) (4.12.2)\n",
            "Requirement already satisfied: wrapt<2,>=1.10 in /usr/local/lib/python3.10/dist-packages (from deprecated<2.0,>=1.2.4->synapseclient) (1.16.0)\n",
            "Requirement already satisfied: anyio in /usr/local/lib/python3.10/dist-packages (from httpx~=0.27.0->synapseclient) (3.7.1)\n",
            "Requirement already satisfied: certifi in /usr/local/lib/python3.10/dist-packages (from httpx~=0.27.0->synapseclient) (2024.8.30)\n",
            "Requirement already satisfied: httpcore==1.* in /usr/local/lib/python3.10/dist-packages (from httpx~=0.27.0->synapseclient) (1.0.6)\n",
            "Requirement already satisfied: idna in /usr/local/lib/python3.10/dist-packages (from httpx~=0.27.0->synapseclient) (3.10)\n",
            "Requirement already satisfied: sniffio in /usr/local/lib/python3.10/dist-packages (from httpx~=0.27.0->synapseclient) (1.3.1)\n",
            "Requirement already satisfied: h11<0.15,>=0.13 in /usr/local/lib/python3.10/dist-packages (from httpcore==1.*->httpx~=0.27.0->synapseclient) (0.14.0)\n",
            "Requirement already satisfied: importlib-metadata<=8.5.0,>=6.0 in /usr/local/lib/python3.10/dist-packages (from opentelemetry-api>=1.21.0->synapseclient) (8.5.0)\n",
            "Requirement already satisfied: googleapis-common-protos~=1.52 in /usr/local/lib/python3.10/dist-packages (from opentelemetry-exporter-otlp-proto-http>=1.21.0->synapseclient) (1.65.0)\n",
            "Requirement already satisfied: opentelemetry-exporter-otlp-proto-common==1.28.0 in /usr/local/lib/python3.10/dist-packages (from opentelemetry-exporter-otlp-proto-http>=1.21.0->synapseclient) (1.28.0)\n",
            "Requirement already satisfied: opentelemetry-proto==1.28.0 in /usr/local/lib/python3.10/dist-packages (from opentelemetry-exporter-otlp-proto-http>=1.21.0->synapseclient) (1.28.0)\n",
            "Requirement already satisfied: protobuf<6.0,>=5.0 in /usr/local/lib/python3.10/dist-packages (from opentelemetry-proto==1.28.0->opentelemetry-exporter-otlp-proto-http>=1.21.0->synapseclient) (5.28.3)\n",
            "Requirement already satisfied: opentelemetry-instrumentation==0.49b0 in /usr/local/lib/python3.10/dist-packages (from opentelemetry-instrumentation-httpx>=0.48b0->synapseclient) (0.49b0)\n",
            "Requirement already satisfied: opentelemetry-semantic-conventions==0.49b0 in /usr/local/lib/python3.10/dist-packages (from opentelemetry-instrumentation-httpx>=0.48b0->synapseclient) (0.49b0)\n",
            "Requirement already satisfied: opentelemetry-util-http==0.49b0 in /usr/local/lib/python3.10/dist-packages (from opentelemetry-instrumentation-httpx>=0.48b0->synapseclient) (0.49b0)\n",
            "Requirement already satisfied: packaging>=18.0 in /usr/local/lib/python3.10/dist-packages (from opentelemetry-instrumentation==0.49b0->opentelemetry-instrumentation-httpx>=0.48b0->synapseclient) (24.1)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/dist-packages (from requests<3.0,>=2.22.0->synapseclient) (3.4.0)\n",
            "Requirement already satisfied: zipp>=3.20 in /usr/local/lib/python3.10/dist-packages (from importlib-metadata<=8.5.0,>=6.0->opentelemetry-api>=1.21.0->synapseclient) (3.20.2)\n",
            "Requirement already satisfied: exceptiongroup in /usr/local/lib/python3.10/dist-packages (from anyio->httpx~=0.27.0->synapseclient) (1.2.2)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import re, io, sys"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 17
        },
        "id": "_qEZe_kiZYvd",
        "executionInfo": {
          "status": "ok",
          "timestamp": 1731013861673,
          "user_tz": 300,
          "elapsed": 140,
          "user": {
            "displayName": "",
            "userId": ""
          }
        },
        "outputId": "2db104e6-d2d2-466d-e520-e9e646fce3f0"
      },
      "id": "_qEZe_kiZYvd",
      "execution_count": 58,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from IPython.display import HTML, Markdown\n",
        "\n",
        "def set_css_in_cell_output(unused):\n",
        "  display(HTML(\"\"\"\"\"\"))\n",
        "\n",
        "get_ipython().events.register('pre_run_cell', set_css_in_cell_output)\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 17
        },
        "id": "Hxh9c5cjVU8U",
        "executionInfo": {
          "status": "ok",
          "timestamp": 1731013863104,
          "user_tz": 300,
          "elapsed": 110,
          "user": {
            "displayName": "",
            "userId": ""
          }
        },
        "outputId": "5e59e317-c39c-4242-bd5a-e5373780965b"
      },
      "id": "Hxh9c5cjVU8U",
      "execution_count": 59,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import base64\n",
        "import vertexai\n",
        "from vertexai.generative_models import GenerativeModel, Part, SafetySetting"
      ],
      "metadata": {
        "id": "Te7k0F13EICY",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 17
        },
        "executionInfo": {
          "status": "ok",
          "timestamp": 1731013864430,
          "user_tz": 300,
          "elapsed": 137,
          "user": {
            "displayName": "",
            "userId": ""
          }
        },
        "outputId": "4222ef79-bb1a-476f-e381-47e3fedb010f"
      },
      "id": "Te7k0F13EICY",
      "execution_count": 60,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Synapse client for data access\n",
        "import synapseclient\n",
        "from synapseclient import Synapse\n",
        "\n",
        "# Google BigQuery and data handling\n",
        "import pandas as pd\n",
        "from google.cloud import bigquery\n",
        "import pandas_gbq  # To read BigQuery data into pandas DataFrames\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 17
        },
        "id": "aBo6yI-pBUUd",
        "executionInfo": {
          "status": "ok",
          "timestamp": 1731013865802,
          "user_tz": 300,
          "elapsed": 181,
          "user": {
            "displayName": "",
            "userId": ""
          }
        },
        "outputId": "e0822c34-b3c2-4611-9910-1aec281034fc"
      },
      "id": "aBo6yI-pBUUd",
      "execution_count": 61,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Token for accessing Synapse API"
      ],
      "metadata": {
        "id": "83E8zNQaThja"
      },
      "id": "83E8zNQaThja"
    },
    {
      "cell_type": "code",
      "source": [
        "token1=\"eyJ0eXAiOiJKV1QiLCJraWQiOiJXN05OOldMSlQ6SjVSSzpMN1RMOlQ3TDc6M1ZYNjpKRU9VOjY0NFI6VTNJWDo1S1oyOjdaQ0s6RlBUSCIsImFsZyI6IlJTMjU2In0.eyJhY2Nlc3MiOnsic2NvcGUiOlsidmlldyIsImRvd25sb2FkIiwibW9kaWZ5Il0sIm9pZGNfY2xhaW1zIjp7fX0sInRva2VuX3R5cGUiOiJQRVJTT05BTF9BQ0NFU1NfVE9LRU4iLCJpc3MiOiJodHRwczovL3JlcG8tcHJvZC5wcm9kLnNhZ2ViYXNlLm9yZy9hdXRoL3YxIiwiYXVkIjoiMCIsIm5iZiI6MTczMDkyNjQ5MSwiaWF0IjoxNzMwOTI2NDkxLCJqdGkiOiIxMzUxOSIsInN1YiI6IjM1MDUyODEifQ.EtYAIYe5ETCKZ_MwMgrnig1ygkeSMONvY9_a8kbfDdbhea5pn8dlRmGUKgV-llPzZqQQlijrITaoNclatU7EUpW_FUVPzH573pcGzOi2SfEwM7N-_dYddd5OgvUnLwcDmAFEAbYSEOtsm-h4CSfrZJaflhP08nSTzWEw30AYRvW28H-1KX_xEBDkOfZC1xyKDStcDdH3-02Tu-hP5POuUqcG2MKmRFco55xtnZ4vg98Izp-4QX-DbafqTO04DIsopYlm8SmCto_4ZoVdzJZNq3ZugcxN_6lb6CnEG6WWh81xIur8Fe9af8ghCTtFDrAENONx7bRlzeA_hsoLstv49w\"\n",
        "syn = synapseclient.login(authToken=token1)"
      ],
      "metadata": {
        "id": "EJAiYKqkOkf5",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 86
        },
        "executionInfo": {
          "status": "ok",
          "timestamp": 1731015015377,
          "user_tz": 300,
          "elapsed": 1100,
          "user": {
            "displayName": "",
            "userId": ""
          }
        },
        "outputId": "d8eb2a79-16c8-40fa-ed77-1f9c65dbd143"
      },
      "id": "EJAiYKqkOkf5",
      "execution_count": 88,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Welcome, krithika.b!\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "INFO:synapseclient_default:Welcome, krithika.b!\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Mounting google drive to see if we can access any data from there"
      ],
      "metadata": {
        "id": "74X_nNQfTb2C"
      },
      "id": "74X_nNQfTb2C"
    },
    {
      "cell_type": "code",
      "source": [
        "# Specify the path to the uploaded file\n",
        "metadata_df = pd.read_csv('/content/HTAN_ST_metadata_ad.tsv', sep='\\t')  # Use sep='\\t' for TSV files\n",
        "\n",
        "# Display the dataframe\n",
        "metadata_df.head()\n",
        "\n"
      ],
      "metadata": {
        "id": "9PV6yTibS8Od",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 206
        },
        "executionInfo": {
          "status": "ok",
          "timestamp": 1731013869260,
          "user_tz": 300,
          "elapsed": 138,
          "user": {
            "displayName": "",
            "userId": ""
          }
        },
        "outputId": "f0e14d57-b154-4474-95e9-555e79fa55e6"
      },
      "id": "9PV6yTibS8Od",
      "execution_count": 62,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                    Attribute:  \\\n",
              "0                    Filename:   \n",
              "1                      Run ID:   \n",
              "2                 File Format:   \n",
              "3  HTAN Parent Biospecimen ID:   \n",
              "4           HTAN Data File ID:   \n",
              "\n",
              "                                         Description  \n",
              "0                                     Name of a file  \n",
              "1  A unique identifier for this individual run (t...  \n",
              "2  Format of a file (e.g. txt, csv, fastq, bam, e...  \n",
              "3  HTAN Biospecimen Identifier (eg HTANx_yyy_zzz)...  \n",
              "4  Self-identifier for this data file - HTAN ID o...  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-66bbf79c-fa03-4a19-9d39-5cc9eb0711e6\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Attribute:</th>\n",
              "      <th>Description</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>Filename:</td>\n",
              "      <td>Name of a file</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Run ID:</td>\n",
              "      <td>A unique identifier for this individual run (t...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>File Format:</td>\n",
              "      <td>Format of a file (e.g. txt, csv, fastq, bam, e...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>HTAN Parent Biospecimen ID:</td>\n",
              "      <td>HTAN Biospecimen Identifier (eg HTANx_yyy_zzz)...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>HTAN Data File ID:</td>\n",
              "      <td>Self-identifier for this data file - HTAN ID o...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-66bbf79c-fa03-4a19-9d39-5cc9eb0711e6')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-66bbf79c-fa03-4a19-9d39-5cc9eb0711e6 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-66bbf79c-fa03-4a19-9d39-5cc9eb0711e6');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-9d0dd7d6-415a-427e-bc45-7d31998da3fd\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-9d0dd7d6-415a-427e-bc45-7d31998da3fd')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-9d0dd7d6-415a-427e-bc45-7d31998da3fd button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ]
          },
          "metadata": {},
          "execution_count": 62
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "PROJECT_ID = \"isb-cgc-external-004\"\n",
        "LOCATION = \"us-central1\"\n",
        "\n",
        "import vertexai\n",
        "vertexai.init(project=PROJECT_ID, location=LOCATION)"
      ],
      "metadata": {
        "id": "d9kb9xRPEX1N",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 17
        },
        "executionInfo": {
          "status": "ok",
          "timestamp": 1731013871482,
          "user_tz": 300,
          "elapsed": 156,
          "user": {
            "displayName": "",
            "userId": ""
          }
        },
        "outputId": "62075434-167d-41c9-f153-e3f22e36fb59"
      },
      "id": "d9kb9xRPEX1N",
      "execution_count": 63,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from vertexai.preview import reasoning_engines"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 17
        },
        "id": "X0c1Y09iYJzD",
        "executionInfo": {
          "status": "ok",
          "timestamp": 1731013873015,
          "user_tz": 300,
          "elapsed": 152,
          "user": {
            "displayName": "",
            "userId": ""
          }
        },
        "outputId": "ce642ddb-699a-40f2-d985-8742750f2917"
      },
      "id": "X0c1Y09iYJzD",
      "execution_count": 64,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "generation_config = {\n",
        "    \"max_output_tokens\": 8192,\n",
        "    \"temperature\": 1,\n",
        "    \"top_p\": 0.5,\n",
        "}\n",
        "\n",
        "safety_settings = [\n",
        "    SafetySetting(\n",
        "        category=SafetySetting.HarmCategory.HARM_CATEGORY_HATE_SPEECH,\n",
        "        threshold=SafetySetting.HarmBlockThreshold.OFF\n",
        "    ),\n",
        "    SafetySetting(\n",
        "        category=SafetySetting.HarmCategory.HARM_CATEGORY_DANGEROUS_CONTENT,\n",
        "        threshold=SafetySetting.HarmBlockThreshold.OFF\n",
        "    ),\n",
        "    SafetySetting(\n",
        "        category=SafetySetting.HarmCategory.HARM_CATEGORY_SEXUALLY_EXPLICIT,\n",
        "        threshold=SafetySetting.HarmBlockThreshold.OFF\n",
        "    ),\n",
        "    SafetySetting(\n",
        "        category=SafetySetting.HarmCategory.HARM_CATEGORY_HARASSMENT,\n",
        "        threshold=SafetySetting.HarmBlockThreshold.OFF\n",
        "    ),\n",
        "]\n",
        "\n",
        "def generate(question = \"\"\"What can you do?\"\"\"):\n",
        "\n",
        "    ## General prompt\n",
        "    prompt = f\"\"\"\n",
        "    You are a bioinformatics expert coder using the Human Tumor Network Atlas (HTAN)\n",
        "    datasets via Google Cloud and Jupyter. You can write Python code to answer\n",
        "    questions by writing BigQuery queries, pandas, scanpy, and squidpy code.\n",
        "    When a user asks a question, identify whether you need to write any code to\n",
        "    answer.\n",
        "    Very important: !! For any general questions and conversations,\n",
        "    you can answer with a conversation. For any questions that needs the code,\n",
        "    just write and return the Python code enclosed in ```python``` tags so that\n",
        "    the user can regex extract the code easily.\n",
        "    You do not have to access any data. You are an expert coder, just write the\n",
        "    code. When you're asked to load a dataset, use BigQuery to load it.\n",
        "\n",
        "    Include all the necessary import packages to run the code. If using any\n",
        "    Google Cloud services, you can use:\n",
        "    project_id = \"isb-cgc-external-004\"\n",
        "\n",
        "    Here are some examples:\n",
        "\n",
        "    Question: Can you load the cells from the HTAN biospecimen HTA7_1_3?\n",
        "    Answer:\n",
        "    '''\n",
        "    query = '''\n",
        "    WITH cells AS (\n",
        "      SELECT  CellID, X_centroid, Y_centroid,\n",
        "      FROM `isb-cgc-bq.HTAN.imaging_level4_HMS_mel_mask_current`\n",
        "      WHERE HTAN_Biospecimen_ID = 'HTA7_1_3'\n",
        "    )\n",
        "    SELECT CellID, X_centroid,  Y_centroid\n",
        "    FROM cells\n",
        "    '''\n",
        "    df = pandas_gbq.read_gbq(query, project_id=project_id)\n",
        "    '''\n",
        "\n",
        "\n",
        "    \"\"\"\n",
        "\n",
        "\n",
        "\n",
        "    #### Synapse data download : Example 1\n",
        "\n",
        "    prompt += f\"\"\"\n",
        "    Question: Can you load the entityId of all data under HTAN.10xvisium_spatialtranscriptomics_scRNAseq_level4_metadata_current\n",
        "    where the File_Format is hdf5?\n",
        "\n",
        "    Answer:\n",
        "    '''\n",
        "    import pandas_gbq\n",
        "    from google.cloud import bigquery\n",
        "\n",
        "    project_id = \"isb-cgc-external-004\"\n",
        "\n",
        "    query = '''\n",
        "    SELECT entityId\n",
        "    FROM `isb-cgc-bq.HTAN.10xvisium_spatialtranscriptomics_scRNAseq_level4_metadata_current`\n",
        "    WHERE File_Format = 'hdf5'\n",
        "    '''\n",
        "\n",
        "    df = pandas_gbq.read_gbq(query, project_id=project_id)\n",
        "    '''\n",
        "\n",
        "    \"\"\"\n",
        "\n",
        "    ##  Synapse example 2\n",
        "\n",
        "    prompt += f\"\"\"\n",
        "    Question: Can you download the synapse data 'syn51133602' to /content/datasets/?\n",
        "    Answer:\n",
        "    '''\n",
        "    import synapseclient\n",
        "    syn = synapseclient.login()\n",
        "    entity = syn.get('syn51133602', downloadLocation='/content/datasets')\n",
        "    '''\n",
        "\n",
        "    \"\"\"\n",
        "\n",
        "    ### Spatial example 1\n",
        "\n",
        "    prompt += f\"\"\"\n",
        "    Question: Can you categorize cells within a defined spatial region as either 'Tumor' or 'Other' based on threshold expression levels of specific markers (SOX10_cellRingMask, S100B_cellRingMask, and CD63_cellRingMask)?\n",
        "    Answer:\n",
        "    '''\n",
        "    query = '''\n",
        "      WITH cells AS (\n",
        "      SELECT CellID, X_centroid, Y_centroid,\n",
        "      IF (SOX10_cellRingMask > 3704.5 AND (S100B_cellRingMask > 7589.48 OR CD63_cellRingMask > 570.68),\n",
        "      'Tumor', 'Other') AS celltype\n",
        "      FROM `isb-cgc-bq.HTAN.imaging_level4_HMS_mel_mask_current`\n",
        "      WHERE HTAN_Biospecimen_ID = 'HTA7_1_3')\n",
        "\n",
        "      SELECT CellID, X_centroid, Y_centroid, celltype\n",
        "      FROM cells\n",
        "      WHERE X_centroid > 23076.9 AND X_centroid < 30384.6\n",
        "      AND Y_centroid > 9615.3 AND Y_centroid < 15000\n",
        "      '''\n",
        "\n",
        "    \"\"\"\n",
        "\n",
        "    ### Spatial example 2\n",
        "    prompt += f\"\"\"\n",
        "    Question: Can you classify cells as 'Tumor' or 'Other', convert their pixel coordinates to geospatial points, and calculate distances between cell pairs that are within a 20-micrometer proximity threshold\n",
        "    Answer:\n",
        "    query = '''\n",
        "      WITH geodat AS (\n",
        "      SELECT CellID, X_centroid, Y_centroid,\n",
        "      IF (SOX10_cellRingMask > 3704.5 AND (S100B_cellRingMask > 7589.48 OR CD63_cellRingMask > 570.68),\n",
        "      'Tumor', 'Other') AS celltype,\n",
        "      ST_GeogPoint(X_centroid / 368570, Y_centroid / 368570) AS p\n",
        "      FROM `isb-cgc-bq.HTAN.imaging_level4_HMS_mel_mask_current`\n",
        "      WHERE HTAN_Biospecimen_ID = 'HTA7_1_3'\n",
        "      )\n",
        "    SELECT t1.CellID, t1.X_centroid, t1.Y_centroid, t1.p, t1.celltype,\n",
        "    t2.CellID AS CellID_1, t2.X_centroid AS X_centroid_1, t2.Y_centroid AS Y_centroid_1, t2.p AS p_1, t2.celltype AS celltype_1,\n",
        "    ST_Distance(t1.p, t2.p) AS Distance\n",
        "    FROM geodat AS t1\n",
        "    JOIN geodat AS t2\n",
        "    ON ST_DWithin(t1.p, t2.p, 9.29324770787722)\n",
        "    '''\n",
        "\n",
        "\n",
        "    \"\"\"\n",
        "\n",
        "    ### Spatial example 3\n",
        "    prompt += f\"\"\"\n",
        "    Question: For each tumor cell within the specified spatial region, calculate the number of neighboring tumor cells\n",
        "    Answer:\n",
        "    query = '''\n",
        "      WITH cellp AS (\n",
        "      SELECT CellID, celltype, CellID_1, celltype_1\n",
        "      FROM `isb-cgc-bq.temp15432.Melanoma_CyCIF_HTA7_1_3_points_within_20um`\n",
        "      WHERE X_centroid > 23076.9 AND X_centroid < 30384.6\n",
        "      AND Y_centroid > 9615.3 AND Y_centroid < 15000)\n",
        "\n",
        "      SELECT CellID, COUNTIF(celltype_1 = 'Tumor') - 1 AS N_Tumor_Cells\n",
        "      FROM cellp\n",
        "      WHERE celltype = 'Tumor'\n",
        "      GROUP BY CellID\n",
        "    '''\n",
        "\n",
        "    \"\"\"\n",
        "\n",
        "\n",
        "    ## Metadata\n",
        "\n",
        "    # Create a prompt with the file content included\n",
        "    prompt += f\"\"\"\n",
        "    The HTAN metadata contains a subset of these type of attributes. Also included\n",
        "    is a description of these attributes. Internalize this information and use it\n",
        "    to answer any queries related to metadata.\n",
        "\n",
        "    Metadata Description:\n",
        "    {metadata_df}\n",
        "\n",
        "    Question: Which attributes describe the spatial location of cells?\n",
        "    Answer:\n",
        "\n",
        "    \"\"\"\n",
        "\n",
        "\n",
        "\n",
        "    ## ScRNAseq example 1\n",
        "    prompt += f\"\"\"\n",
        "    Question: What are the counts of unique cells, sex groupings, samples,\n",
        "    cell types, and therapies by development stage in the MSK scRNAseq dataset?\n",
        "    Answer:\n",
        "    '''\n",
        "    query = '''\n",
        "    SELECT\n",
        "      development_stage,\n",
        "      count(distinct(iObs)) AS Number_Cells,\n",
        "      count(distinct(sex)) AS Unique_Sex_Grouping,\n",
        "      count(distinct(donor_id)) AS Number_Samples,\n",
        "      count(distinct(cell_type)) AS Number_Cell_Types,\n",
        "      count(distinct(treatment)) AS Number_Therapies\n",
        "    FROM\n",
        "      `isb-cgc-bq.HTAN.scRNAseq_MSK_SCLC_combined_samples_current`\n",
        "    GROUP BY development_stage\n",
        "    ORDER BY Number_Samples DESC\n",
        "    '''\n",
        "    df = pandas_gbq.read_gbq(query, project_id=project_id)\n",
        "    '''\n",
        "\n",
        "    \"\"\"\n",
        "\n",
        "    ## ScRNAseq example 2\n",
        "    prompt += f\"\"\"\n",
        "    Question: How many unique cell types, sex groupings, cells, and samples\n",
        "    are present in the a specific human stage of development in the MSK\n",
        "    scRNAseq dataset (e.g. 74-year-old)?\n",
        "    Answer:\n",
        "    '''\n",
        "    query = '''\n",
        "    SELECT\n",
        "      cell_type,\n",
        "      count(distinct(sex)) AS Unique_Sex_Grouping,\n",
        "      count(distinct(iObs)) AS Number_Cells,\n",
        "      count(distinct(donor_id)) AS Number_Samples\n",
        "    FROM\n",
        "      `isb-cgc-bq.HTAN.scRNAseq_MSK_SCLC_combined_samples_current`\n",
        "    WHERE\n",
        "      development_stage = '74-year-old human stage'\n",
        "    GROUP BY cell_type\n",
        "    '''\n",
        "    df = pandas_gbq.read_gbq(query, project_id=project_id)\n",
        "    '''\n",
        "\n",
        "    \"\"\"\n",
        "\n",
        "    ## ScRNAseq example 3\n",
        "    prompt += f\"\"\"\n",
        "    Question: How many genes and cells are associated with each sex and\n",
        "    cell type in the MSK scRNAseq dataset for an individual\n",
        "    (e.g. a 74-year-old human stage)?\n",
        "    Answer:\n",
        "    '''\n",
        "    query = '''\n",
        "    SELECT\n",
        "      sex,\n",
        "      Cell_Type,\n",
        "      count(distinct(feature_name)) AS Number_Genes,\n",
        "      count(distinct(iObs)) AS Number_Cells\n",
        "    FROM\n",
        "      `isb-cgc-bq.HTAN.scRNAseq_MSK_SCLC_combined_samples_current`\n",
        "    WHERE development_stage = '74-year-old human stage'\n",
        "    GROUP BY sex, Cell_Type\n",
        "    ORDER BY Cell_Type DESC\n",
        "    '''\n",
        "    df = pandas_gbq.read_gbq(query, project_id=project_id)\n",
        "    '''\n",
        "\n",
        "    \"\"\"\n",
        "\n",
        "    ## ScRNAseq example 4\n",
        "    prompt += f\"\"\"\n",
        "    Question: How many genes and cells are there in each Seurat Cluster\n",
        "    for males and females of the 'epithelial cell' type in the specific\n",
        "    human stage (here 74-year-old)?\n",
        "    Answer:\n",
        "    '''\n",
        "    query = '''\n",
        "    SELECT\n",
        "      sex,\n",
        "      clusters,\n",
        "      Cell_Type,\n",
        "      count(distinct(feature_name)) AS Number_Genes,\n",
        "      count(distinct(iObs)) AS Number_Cells\n",
        "    FROM\n",
        "      `isb-cgc-bq.HTAN.scRNAseq_MSK_SCLC_combined_samples_current`\n",
        "    WHERE development_stage = '74-year-old human stage' AND Cell_Type = 'epithelial cell'\n",
        "    GROUP BY sex, clusters, Cell_Type\n",
        "    ORDER BY clusters ASC\n",
        "    '''\n",
        "    df = pandas_gbq.read_gbq(query, project_id=project_id)\n",
        "    '''\n",
        "\n",
        "    \"\"\"\n",
        "\n",
        "   ## ScRNAseq example 5\n",
        "    prompt += f\"\"\"\n",
        "    Question: How do the average expression values for genes differ between\n",
        "    male and female epithelial cells in a specific cluster, and which genes\n",
        "    show the greatest differences (here cluster 41 of the 74-year-old\n",
        "    human stage)?\n",
        "    Answer:\n",
        "    '''\n",
        "    query = '''\n",
        "    SELECT\n",
        "      A.feature_name,\n",
        "      A.avg_counts_clust10 AS female_avg_counts,\n",
        "      B.avg_counts_clust10 AS male_avg_counts,\n",
        "      A.avg_counts_clust10 - B.avg_counts_clust10 AS mean_diff\n",
        "    FROM (\n",
        "      SELECT\n",
        "        feature_name,\n",
        "        AVG(X_value) AS avg_counts_clust10\n",
        "      FROM\n",
        "        `isb-cgc-bq.HTAN.scRNAseq_MSK_SCLC_combined_samples_current`\n",
        "      WHERE development_stage = '74-year-old human stage' AND Cell_Type = 'epithelial cell' AND clusters = '41' AND sex = 'female'\n",
        "      GROUP BY feature_name\n",
        "    ) AS A\n",
        "    INNER JOIN (\n",
        "      SELECT\n",
        "        feature_name,\n",
        "        AVG(X_value) AS avg_counts_clust10\n",
        "      FROM\n",
        "        `isb-cgc-bq.HTAN.scRNAseq_MSK_SCLC_combined_samples_current`\n",
        "      WHERE development_stage = '74-year-old human stage' AND Cell_Type = 'epithelial cell' AND clusters = '41' AND sex = 'male'\n",
        "      GROUP BY feature_name\n",
        "    ) AS B\n",
        "    ON A.feature_name = B.feature_name\n",
        "    ORDER BY mean_diff DESC\n",
        "    '''\n",
        "    df = pandas_gbq.read_gbq(query, project_id=project_id)\n",
        "    '''\n",
        "\n",
        "    \"\"\"\n",
        "\n",
        "\n",
        "    # Add the dynamic user question and the placeholder for the answer\n",
        "    prompt += f\"\"\"\n",
        "    Question: {question}\n",
        "    Answer:\n",
        "    \"\"\"\n",
        "\n",
        "\n",
        "    # Model information\n",
        "    vertexai.init(project=\"isb-cgc-external-004\", location=\"us-central1\")\n",
        "    model = GenerativeModel(\n",
        "        \"gemini-1.5-flash-002\",\n",
        "    )\n",
        "\n",
        "    # Responses\n",
        "    responses = model.generate_content(\n",
        "        [prompt],\n",
        "        generation_config=generation_config,\n",
        "        safety_settings=safety_settings,\n",
        "        stream=False,\n",
        "        # tools='code_execution'\n",
        "    )\n",
        "\n",
        "    # for response in responses:\n",
        "    #     print(response.text, end=\"\")\n",
        "\n",
        "    pattern = r'```python\\n(.*?)\\n```'\n",
        "\n",
        "    # Extract the code\n",
        "    match = re.search(pattern, responses.text, re.DOTALL)\n",
        "\n",
        "    if match:\n",
        "        python_code = match.group(1).strip()\n",
        "        print(python_code)\n",
        "        return(python_code)\n",
        "\n",
        "    else:\n",
        "        print(responses)\n",
        "\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 17
        },
        "id": "qhnOrkXrYMxQ",
        "executionInfo": {
          "status": "ok",
          "timestamp": 1731014374520,
          "user_tz": 300,
          "elapsed": 169,
          "user": {
            "displayName": "",
            "userId": ""
          }
        },
        "outputId": "be3289cd-5b74-4275-a3ad-7b5a20d1687f"
      },
      "id": "qhnOrkXrYMxQ",
      "execution_count": 76,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Test 1\n",
        "code=generate('Who are you and what can you do?')\n",
        "print(code)"
      ],
      "metadata": {
        "id": "GO8pCZGoZRAi"
      },
      "id": "GO8pCZGoZRAi",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Test 2\n",
        "code=generate(\"Can you load the cells from the HTAN biospecimen HTA7_1_3?\")\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 277
        },
        "id": "UeL84jhjYxpm",
        "executionInfo": {
          "status": "ok",
          "timestamp": 1731013888159,
          "user_tz": 300,
          "elapsed": 899,
          "user": {
            "displayName": "",
            "userId": ""
          }
        },
        "outputId": "5f8f62d9-c764-41ae-8d54-d060aaee35b1"
      },
      "id": "UeL84jhjYxpm",
      "execution_count": 67,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "import pandas_gbq\n",
            "\n",
            "project_id = \"isb-cgc-external-004\"\n",
            "\n",
            "query = \"\"\"\n",
            "WITH cells AS (\n",
            "  SELECT  CellID, X_centroid, Y_centroid,\n",
            "  FROM `isb-cgc-bq.HTAN.imaging_level4_HMS_mel_mask_current`\n",
            "  WHERE HTAN_Biospecimen_ID = 'HTA7_1_3'\n",
            ")\n",
            "SELECT CellID, X_centroid,  Y_centroid\n",
            "FROM cells\n",
            "\"\"\"\n",
            "\n",
            "df = pandas_gbq.read_gbq(query, project_id=project_id)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "exec(code)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 52
        },
        "id": "dBZS1gGQZji_",
        "executionInfo": {
          "status": "ok",
          "timestamp": 1731013935614,
          "user_tz": 300,
          "elapsed": 44523,
          "user": {
            "displayName": "",
            "userId": ""
          }
        },
        "outputId": "98f26876-9117-4f38-f064-2454ee082207"
      },
      "id": "dBZS1gGQZji_",
      "execution_count": 68,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "WARNING: google.colab.auth.authenticate_user() is not supported in Colab Enterprise.\n",
            "Downloading: 100%|\u001b[32m██████████\u001b[0m|\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "print(df)"
      ],
      "metadata": {
        "id": "p6zoT7yqc8Nc",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 260
        },
        "executionInfo": {
          "status": "ok",
          "timestamp": 1731006723059,
          "user_tz": 300,
          "elapsed": 141,
          "user": {
            "displayName": "",
            "userId": ""
          }
        },
        "outputId": "2d37c8a7-d216-4391-a975-39bd810eef6e"
      },
      "id": "p6zoT7yqc8Nc",
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "          CellID    X_centroid    Y_centroid\n",
            "0         414954  12234.156425  11496.055866\n",
            "1         414965  17100.977778  11491.444444\n",
            "2         414982  20372.921053  11492.671053\n",
            "3         415004  30851.258170  11497.666667\n",
            "4         415185  10789.783784  11496.283784\n",
            "...          ...           ...           ...\n",
            "1110580  1099321  29743.450000  23012.790000\n",
            "1110581  1099373   8805.850649  23025.246753\n",
            "1110582  1099466  24451.210526  23035.491228\n",
            "1110583  1099530  29740.808511  23049.994681\n",
            "1110584  1099567  10085.244186  23054.709302\n",
            "\n",
            "[1110585 rows x 3 columns]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Test 3\n",
        "code=generate(\"Can you load the cells from the HTAN biospecimen HTA7_8_2?\")\n",
        "exec(code)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 295
        },
        "id": "StSCvi1rY9c7",
        "executionInfo": {
          "status": "ok",
          "timestamp": 1731014011606,
          "user_tz": 300,
          "elapsed": 19944,
          "user": {
            "displayName": "",
            "userId": ""
          }
        },
        "outputId": "26d8bd19-aafa-466a-937b-009e050a5ddc"
      },
      "id": "StSCvi1rY9c7",
      "execution_count": 70,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "import pandas_gbq\n",
            "\n",
            "project_id = \"isb-cgc-external-004\"\n",
            "\n",
            "query = \"\"\"\n",
            "WITH cells AS (\n",
            "  SELECT  CellID, X_centroid, Y_centroid,\n",
            "  FROM `isb-cgc-bq.HTAN.imaging_level4_HMS_mel_mask_current`\n",
            "  WHERE HTAN_Biospecimen_ID = 'HTA7_8_2'\n",
            ")\n",
            "SELECT CellID, X_centroid,  Y_centroid\n",
            "FROM cells\n",
            "\"\"\"\n",
            "\n",
            "df = pandas_gbq.read_gbq(query, project_id=project_id)\n",
            "Downloading: 100%|\u001b[32m██████████\u001b[0m|\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "print(df)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 260
        },
        "id": "reX7bQDVcyou",
        "executionInfo": {
          "status": "ok",
          "timestamp": 1731014011606,
          "user_tz": 300,
          "elapsed": 2,
          "user": {
            "displayName": "",
            "userId": ""
          }
        },
        "outputId": "8b7c14ee-0f7b-430e-e677-3f8ffc87f356"
      },
      "id": "reX7bQDVcyou",
      "execution_count": 71,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "        CellID    X_centroid    Y_centroid\n",
            "0       271194  28345.181303  22522.594901\n",
            "1       271210   8021.079452  22523.334247\n",
            "2       271276   6641.683544  22525.746835\n",
            "3       271353   5904.687980  22531.846547\n",
            "4       271388   4592.990741  22528.296296\n",
            "...        ...           ...           ...\n",
            "352915  270827   4276.356322  22506.913793\n",
            "352916  270863  10172.643750  22508.637500\n",
            "352917  270889  26049.331034  22509.006897\n",
            "352918  270893   5114.777778  22508.931624\n",
            "352919  271067  11497.290179  22517.410714\n",
            "\n",
            "[352920 rows x 3 columns]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "## Test 3\n",
        "code=generate(\"What kind of information is captured in the HTAN spatial transcriptomics metadata file?\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 332
        },
        "id": "cErGMPtYdIT4",
        "executionInfo": {
          "status": "ok",
          "timestamp": 1731014380132,
          "user_tz": 300,
          "elapsed": 1073,
          "user": {
            "displayName": "",
            "userId": ""
          }
        },
        "outputId": "fd8d7cb9-8449-4fdc-f16d-895c1880fe95"
      },
      "id": "cErGMPtYdIT4",
      "execution_count": 77,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "candidates {\n",
            "  content {\n",
            "    role: \"model\"\n",
            "    parts {\n",
            "      text: \"The HTAN spatial transcriptomics metadata file contains information about the spatial location of cells, including attributes such as `Spatial Barcode Length`, `UMI Barcode Offset`, `UMI Barcode Length`, and potentially coordinates if available in the specific dataset.  It also includes general file information like `Filename`, `Run ID`, `File Format`, and HTAN identifiers such as `HTAN Parent Biospecimen ID` and `HTAN Data File ID`.  The exact attributes will vary depending on the specific dataset.\\n\"\n",
            "    }\n",
            "  }\n",
            "  finish_reason: STOP\n",
            "  avg_logprobs: -0.10189473395254098\n",
            "}\n",
            "usage_metadata {\n",
            "  prompt_token_count: 3155\n",
            "  candidates_token_count: 102\n",
            "  total_token_count: 3257\n",
            "}\n",
            "model_version: \"gemini-1.5-flash-002\"\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "print(code)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 34
        },
        "id": "pApmXcqDf2UU",
        "executionInfo": {
          "status": "ok",
          "timestamp": 1731014392985,
          "user_tz": 300,
          "elapsed": 124,
          "user": {
            "displayName": "",
            "userId": ""
          }
        },
        "outputId": "f1e1dd9f-8326-40a5-fcb4-774980b90113"
      },
      "id": "pApmXcqDf2UU",
      "execution_count": 78,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "None\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "## Test 4\n",
        "code=generate(\"What kind of information is captured in the Synapse database?\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 332
        },
        "id": "5OflnZZnusYl",
        "executionInfo": {
          "status": "ok",
          "timestamp": 1731014404601,
          "user_tz": 300,
          "elapsed": 1248,
          "user": {
            "displayName": "",
            "userId": ""
          }
        },
        "outputId": "e6c5add9-6fb2-4646-bc58-611cf36d600f"
      },
      "id": "5OflnZZnusYl",
      "execution_count": 79,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "candidates {\n",
            "  content {\n",
            "    role: \"model\"\n",
            "    parts {\n",
            "      text: \"The Synapse database contains a wide variety of data types relevant to biomedical research.  This includes, but is not limited to, genomic data (e.g., gene expression, mutations), imaging data (e.g., microscopy images, histology slides), clinical data (e.g., patient demographics, treatment information), and metadata describing the experimental procedures and data processing steps.  The specific contents vary depending on the individual datasets stored within Synapse.  To get a precise answer about the contents of a specific Synapse ID, you\\'ll need to provide that ID.\\n\"\n",
            "    }\n",
            "  }\n",
            "  finish_reason: STOP\n",
            "  avg_logprobs: -0.17099447908072635\n",
            "}\n",
            "usage_metadata {\n",
            "  prompt_token_count: 3150\n",
            "  candidates_token_count: 116\n",
            "  total_token_count: 3266\n",
            "}\n",
            "model_version: \"gemini-1.5-flash-002\"\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "### Test 6\n",
        "\n",
        "code=generate(\"Labels cells as 'Tumor' or 'Other' based on the expression of certain markers such as CD63_cellRingMask\")\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 312
        },
        "id": "AQJLm42G1PQL",
        "executionInfo": {
          "status": "ok",
          "timestamp": 1731014408113,
          "user_tz": 300,
          "elapsed": 1206,
          "user": {
            "displayName": "",
            "userId": ""
          }
        },
        "outputId": "212b72ce-9b9a-42d0-89e6-a239bb6966ea"
      },
      "id": "AQJLm42G1PQL",
      "execution_count": 80,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "import pandas_gbq\n",
            "from google.cloud import bigquery\n",
            "\n",
            "project_id = \"isb-cgc-external-004\"\n",
            "\n",
            "query = \"\"\"\n",
            "SELECT\n",
            "    CellID,\n",
            "    CASE\n",
            "        WHEN CD63_cellRingMask > 570.68 THEN 'Tumor'\n",
            "        ELSE 'Other'\n",
            "    END AS cell_type\n",
            "FROM\n",
            "    `isb-cgc-bq.HTAN.imaging_level4_HMS_mel_mask_current`\n",
            "\"\"\"\n",
            "\n",
            "df = pandas_gbq.read_gbq(query, project_id=project_id)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "exec(code)\n",
        "print(df)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 277
        },
        "id": "d3ZWv0YH2xoa",
        "executionInfo": {
          "status": "ok",
          "timestamp": 1731014586689,
          "user_tz": 300,
          "elapsed": 174717,
          "user": {
            "displayName": "",
            "userId": ""
          }
        },
        "outputId": "31bd1afd-a26f-4a3f-f530-06fa6e57c77c"
      },
      "id": "d3ZWv0YH2xoa",
      "execution_count": 81,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Downloading: 100%|\u001b[32m██████████\u001b[0m|\n",
            "         CellID cell_type\n",
            "0        271351     Other\n",
            "1        271535     Other\n",
            "2        271733     Other\n",
            "3        272229     Other\n",
            "4        272613     Other\n",
            "...         ...       ...\n",
            "6282743   83619     Other\n",
            "6282744  201761     Other\n",
            "6282745  450282     Other\n",
            "6282746   71082     Other\n",
            "6282747    5520     Other\n",
            "\n",
            "[6282748 rows x 2 columns]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Question 1\n",
        "print(\"Question 1\")\n",
        "code = generate(\"Can you load the cells from the HTAN biospecimen HTA7_1_3?\")\n",
        "exec(code)\n",
        "\n",
        "# Question 2\n",
        "print(\"Question 2\")\n",
        "code = generate(\"Can you load the entityId of all data under HTAN.10xvisium_spatialtranscriptomics_scRNAseq_level4_metadata_current where the File_Format is hdf5?\")\n",
        "exec(code)\n",
        "\n",
        "# Question 4\n",
        "print(\"Question 4\")\n",
        "code = generate(\"Label cells as 'Tumor' or 'Other' based on the expression of certain markers (SOX10_cellRingMask, S100B_cellRingMask, and CD63_cellRingMask). Restricts the results to cells within a specific spatial region.\")\n",
        "exec(code)\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "nyzVWG88_DEq",
        "executionInfo": {
          "status": "ok",
          "timestamp": 1731015014284,
          "user_tz": 300,
          "elapsed": 59998,
          "user": {
            "displayName": "",
            "userId": ""
          }
        },
        "outputId": "3ead35c8-8713-4108-a8fc-8bce0077cd5c"
      },
      "id": "nyzVWG88_DEq",
      "execution_count": 87,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Question 1\n",
            "import pandas_gbq\n",
            "\n",
            "project_id = \"isb-cgc-external-004\"\n",
            "\n",
            "query = \"\"\"\n",
            "WITH cells AS (\n",
            "  SELECT  CellID, X_centroid, Y_centroid,\n",
            "  FROM `isb-cgc-bq.HTAN.imaging_level4_HMS_mel_mask_current`\n",
            "  WHERE HTAN_Biospecimen_ID = 'HTA7_1_3'\n",
            ")\n",
            "SELECT CellID, X_centroid,  Y_centroid\n",
            "FROM cells\n",
            "\"\"\"\n",
            "\n",
            "df = pandas_gbq.read_gbq(query, project_id=project_id)\n",
            "\n",
            "Downloading:   0%|\u001b[32m          \u001b[0m|\u001b[A\n",
            "Downloading:  11%|\u001b[32m█         \u001b[0m|\u001b[A\n",
            "Downloading:  21%|\u001b[32m██        \u001b[0m|\u001b[A\n",
            "Downloading:  32%|\u001b[32m███▏      \u001b[0m|\u001b[A\n",
            "Downloading:  42%|\u001b[32m████▏     \u001b[0m|\u001b[A\n",
            "Downloading:  53%|\u001b[32m█████▎    \u001b[0m|\u001b[A\n",
            "Downloading:  63%|\u001b[32m██████▎   \u001b[0m|\u001b[A\n",
            "Downloading:  74%|\u001b[32m███████▎  \u001b[0m|\u001b[A\n",
            "Downloading:  84%|\u001b[32m████████▍ \u001b[0m|\u001b[A\n",
            "Downloading:  95%|\u001b[32m█████████▍\u001b[0m|\u001b[A\n",
            "Downloading: 100%|\u001b[32m██████████\u001b[0m|\n",
            "Question 2\n",
            "import pandas_gbq\n",
            "from google.cloud import bigquery\n",
            "\n",
            "project_id = \"isb-cgc-external-004\"\n",
            "\n",
            "query = '''\n",
            "SELECT entityId\n",
            "FROM `isb-cgc-bq.HTAN.10xvisium_spatialtranscriptomics_scRNAseq_level4_metadata_current`\n",
            "WHERE File_Format = 'hdf5'\n",
            "'''\n",
            "\n",
            "df = pandas_gbq.read_gbq(query, project_id=project_id)\n",
            "\n",
            "Downloading: 100%|\u001b[32m██████████\u001b[0m|\n",
            "Question 4\n",
            "import pandas_gbq\n",
            "\n",
            "project_id = \"isb-cgc-external-004\"\n",
            "\n",
            "query = \"\"\"\n",
            "WITH cells AS (\n",
            "  SELECT \n",
            "    CellID, \n",
            "    X_centroid, \n",
            "    Y_centroid,\n",
            "    CASE \n",
            "      WHEN SOX10_cellRingMask > 3704.5 AND (S100B_cellRingMask > 7589.48 OR CD63_cellRingMask > 570.68) THEN 'Tumor' \n",
            "      ELSE 'Other' \n",
            "    END AS celltype\n",
            "  FROM `isb-cgc-bq.HTAN.imaging_level4_HMS_mel_mask_current`\n",
            "  WHERE HTAN_Biospecimen_ID = 'HTA7_1_3'\n",
            ")\n",
            "\n",
            "SELECT \n",
            "  CellID, \n",
            "  X_centroid, \n",
            "  Y_centroid, \n",
            "  celltype\n",
            "FROM cells\n",
            "WHERE X_centroid > 23076.9 AND X_centroid < 30384.6\n",
            "  AND Y_centroid > 9615.3 AND Y_centroid < 15000\n",
            "\"\"\"\n",
            "\n",
            "df = pandas_gbq.read_gbq(query, project_id=project_id)\n",
            "\n",
            "Downloading:   0%|\u001b[32m          \u001b[0m|\u001b[A\n",
            "Downloading:  87%|\u001b[32m████████▋ \u001b[0m|\u001b[A\n",
            "Downloading: 100%|\u001b[32m██████████\u001b[0m|\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Question 3\n",
        "code = generate(\"Can you download the synapse data 'syn51133602' to /content/datasets/?\")\n"
      ],
      "metadata": {
        "id": "8uX3ixJch_5i"
      },
      "id": "8uX3ixJch_5i",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Question 5\n",
        "print(\"Question 5\")\n",
        "code = generate(\"Labels cells as 'Tumor' or 'Other', and converts pixel coordinates to geospatial points. Calculate the distance between cell pairs that fall within a 20-micrometer threshold.\")\n",
        "exec(code)\n",
        "\n",
        "# Question 6\n",
        "print(\"Question 6\")\n",
        "code = generate(\"Calculates the number of neighboring tumor cells for each tumor cell within the specified spatial region.\")\n",
        "exec(code)\n",
        "\n",
        "# Question 7\n",
        "print(\"Question 7\")\n",
        "code = generate(\"What are the counts of unique cells, sex groupings, samples, cell types, and therapies by development stage in the MSK scRNAseq dataset?\")\n",
        "exec(code)\n",
        "\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 746
        },
        "id": "PEvndiwLNLmi",
        "outputId": "a76f4b54-08c5-44f3-ee89-4730139dd8c2"
      },
      "id": "PEvndiwLNLmi",
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": []
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Question 5\n",
            "import pandas_gbq\n",
            "from google.cloud import bigquery\n",
            "\n",
            "project_id = \"isb-cgc-external-004\"\n",
            "\n",
            "query = \"\"\"\n",
            "WITH geodat AS (\n",
            "  SELECT\n",
            "    CellID,\n",
            "    X_centroid,\n",
            "    Y_centroid,\n",
            "    IF(\n",
            "      SOX10_cellRingMask > 3704.5\n",
            "      AND (S100B_cellRingMask > 7589.48 OR CD63_cellRingMask > 570.68),\n",
            "      'Tumor',\n",
            "      'Other'\n",
            "    ) AS celltype,\n",
            "    ST_GeogPoint(X_centroid / 368570, Y_centroid / 368570) AS p\n",
            "  FROM\n",
            "    `isb-cgc-bq.HTAN.imaging_level4_HMS_mel_mask_current`\n",
            "  WHERE HTAN_Biospecimen_ID = 'HTA7_1_3'\n",
            ")\n",
            "SELECT\n",
            "  t1.CellID,\n",
            "  t1.X_centroid,\n",
            "  t1.Y_centroid,\n",
            "  t1.p,\n",
            "  t1.celltype,\n",
            "  t2.CellID AS CellID_1,\n",
            "  t2.X_centroid AS X_centroid_1,\n",
            "  t2.Y_centroid AS Y_centroid_1,\n",
            "  t2.p AS p_1,\n",
            "  t2.celltype AS celltype_1,\n",
            "  ST_Distance(t1.p, t2.p) AS Distance\n",
            "FROM\n",
            "  geodat AS t1\n",
            "JOIN geodat AS t2 ON ST_DWithin(t1.p, t2.p, 9.29324770787722)\n",
            "\n",
            "\"\"\"\n",
            "\n",
            "df = pandas_gbq.read_gbq(query, project_id=project_id)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Question 8\n",
        "print(\"Question 8\")\n",
        "code = generate(\"How many unique cell types, sex groupings, cells, and samples are present in a specific human stage of development in the MSK scRNAseq dataset (e.g. 74-year-old)?\")\n",
        "exec(code)\n",
        "\n",
        "# Question 9\n",
        "print(\"Question 9\")\n",
        "code = generate(\"How many genes and cells are associated with each sex and cell type in the MSK scRNAseq dataset for an individual (e.g. a 74-year-old human stage)?\")\n",
        "exec(code)\n",
        "\n",
        "# Question 10\n",
        "print(\"Question 10\")\n",
        "code = generate(\"How many genes and cells are there in each Seurat Cluster for males and females of the 'epithelial cell' type in the specific human stage (here 74-year-old)?\")\n",
        "exec(code)"
      ],
      "metadata": {
        "id": "PfaRR-0xNVNV"
      },
      "id": "PfaRR-0xNVNV",
      "execution_count": null,
      "outputs": []
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "Python 3",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.10.10"
    },
    "colab": {
      "provenance": [],
      "name": "agent1_kb"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}